test/test.cpp

```diff
@@ -146,16 +147,20 @@ BOOST_AUTO_TEST_CASE( oom_flush_dirty_pages ) {
       db.create<book>( [i]( book& b ) { b.a = (int)i; b.b = (int)(i+1); } );
       if (i % 1000 == 0) {
          if (auto res = db.check_memory_and_flush_if_needed()) {
+            // we need to wait some time after clearing the soft-dirty bits from the task's PTEs
+            // for the next read to be accurate (see https://www.kernel.org/doc/Documentation/vm/soft-dirty.txt)
```

Comment:
Are you saying clearing the bits is asynchronous? If that's the case the whole approach seems dangerous.

I was initially worried about that when doing the initial review, but when I looked through the kernel source it _seemed_ like it was synchronous.

Or instead is the problem here in this test that the `oom_score` is only updated at some interval and we need to wait around for that update to occur? If that's the case, I don't think citing that `soft-dirty.txt` refers to the same issue.

Reply:
The issue is not updating the oom_score (in the test we don't care about the value being updated because we set an arbitrarily low threshold).
It seems that if you check for the soft-dirty bits on the pages right after clearing them, you don't get correct results.

The reason I linked this url is:

<img width="453" alt="image" src="https://github.com/AntelopeIO/chainbase/assets/1496907/d354f017-4d69-4630-aca1-080bc2460ffb">

Reply:
It seems that when we don't wait, we see more pages as dirty as we should, so I don't think it is dangerous, except for the fact that we'd flush more pages to disk than strictly needed. But I agree that it is unnerving.

Reply:
It's dangerous because if it's not synchronous you'll miss writes that occur on pages after writing to `clear_refs` but before the dirty bit is cleared asynchronously in the background.

After `save_database_file()` we should have completely fresh pages anyways though right? So why are we seeing seemingly spurious dirty bits in brand new pages? Is it possible the dirty bit is only valid based on some other criteria? (some other bit being set?)

Reply:
> It's dangerous because if it's not synchronous you'll miss writes that occur on pages after writing to clear_refs but before the dirty bit is cleared asynchronously in the background.

If it behaved like that the feature would be utterly useless.

>  Is it possible the dirty bit is only valid based on some other criteria? (some other bit being set?)

No I don't think so because all we need to do is wait a second or two and then we see the correct pages dirty.

Reply:
Why do the kernel unit tests not require waiting?

Reply:
What kernel unit tests are you referring to?

Reply:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/tools/testing/selftests/mm/soft-dirty.c?h=v6.5.7#n127

Reply:
@spoonincode you are correct, there was an issue in the test (now corrected) that made me think that `clear_refs` did not clear the `soft-dirty` bit immediately, but that was an error on my part. So my original comment was wrong, the clearing of the  `soft-dirty` bits is synchronous.